tests/ui/type-inference/generalize-subtyped-variables.rs

```diff
@@ -0,0 +1,14 @@
+// Test that generalize handles variables that have a subtype relation to an
+// already resolved value, but not an equality relation.
+// Found when considering fixes to #117151
+// check-pass
```

Comment:
This test feels a bit :/ currently the way we generalize higher ranked types is wrong when subtyping. 

we currently do the following (using `x` for the infer var in `x: None<_>`):
- emit (and stall on) a `Subtype(x <: y)` obligation 
- the `w = a` assignment sets `x` to `for<'a> fn(&'a ())`, emits a `Subtype(y <: z)` obligation and generalizes `z` to `fn(&'0 ())`.
- we then prove `Subtype(x <: y)` setting `y` to `for<'a> fn(&'a ())` and then prove `Subtype(y <: z)` which is now trivial as both types are rigid.

If we were to prove the `Subtype` obligations before generalizing `z` we'd get an incorrect error, e.g:
```rust
fn main() {
    let mut x = None;
    let y = x;
    let z = Default::default();
    let mut w = (&mut x, z, z);
    let mut temp = None::<for<'a> fn(&'a ())>;
    w.0 = &mut temp;
    w.1 = y;
    w.2 = None::<fn(&'static ())>; //~ ERROR mismatched types
}
```
It feels like this test mostly tests a very specific impl detail of our subtyping algorithm where we do not try to eagerly make progress on previous `Subtype` obligations while relating.

I think the explanation here is somewhat misleading. The reason this works is because generalization **does not** care about subtype relations (apart from the occurs check). I would instead use sth like "Subtyping of inference variables is currently handled via delayed `Subtype` obligations. We do not immediately process them after constraining one of the involved inference variables. We'd otherwise incorrectly constrain the type of `z` as generalization does not handle binders correctly".